src/compiler/scala/tools/nsc/typechecker/TypeDiagnostics.scala

```diff
@@ -698,11 +698,8 @@ trait TypeDiagnostics {
         } else f
       }
       def apply(tree: Tree): Tree = {
-        // Error suppression (in context.warning) would squash some of these warnings.
-        // It is presumed if you are using a -Y option you would really like to hear
-        // the warnings you've requested; thus, use reporter.warning.
         if (settings.warnDeadCode && context.unit.exists && treeOK(tree) && exprOK)
-          reporter.warning(tree.pos, "dead code following this construct")
+          context.warning(tree.pos, "dead code following this construct")
```

Comment:
This change reverts a conscious choice made by @adriaanm to circumvent the context reporter. In an offline discussion we agreed to revisit, so here it is.

IMO it's better to go via the Context for two reasons:
- if the user wanted to silence a warning I think that should take precedence over a general/coarse-grained command line option
- consistency. it is/will be simpler to enforce reporting policies if all code goes through the same gate